### PR TITLE
DUOS-2011 [risk=no] Local Proxy setup for npm start

### PR DESCRIPTION
**Pull request by JVThomas**
Addresses [DUOS-2011](https://broadworkbench.atlassian.net/browse/DUOS-2011), goal is to set up middleware to act as proxy re-routes for external apis.

Only requirements for running this (so far) are...

- Having a `server.crt` and a `server.key` saved to the base level of your local working directory
- Having a `.env.local` file defined (see DUOS messages)
- Having `env: 'local'` in your `config.json`

Proxy middleware will be set up when you run `npm start`

Have you read [Terra's Contributing Guide](https://github.com/DataBiosphere/terra-ui/wiki/Contributor-Guide) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/libs/notificationService.js [hunk 1]
```diff
@@ -15,7 +15,12 @@ export const NotificationService = {
    */
   getBanners: async () => {
     const env = await Config.getEnv();
-    return $.getJSON(gcs + '/' + env + '_' + bannerFileName, (data) => {
+    const hash = await Config.getHash();
+    const url =
+      env === 'local'
+        ? `/broad-duos-banner/${hash}_${bannerFileName}`
+        : gcs + '/' + env + '_' + bannerFileName;
+    return $.getJSON(url, (data) => {
```
Comment (rushtong) [inline src/libs/notificationService.js:23]: Optional ... the fact that this url construction was in `NotificationService` is a bit of an outlier and older code. It would be nice if this were consolidated somewhere else, like `ajax.js` or `config.js`. Definitely not a PR blocker.

File: src/setupProxy.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+const { createProxyMiddleware } = require('http-proxy-middleware');
+const configUrls = require('../public/config.json');
+
+module.exports = function (app) {
```
Comment (rushtong) [inline src/setupProxy.js:31]: I feel like we could use a hefty dose of documentation here, and perhaps in `README.md` as well.

**Review by rushtong — COMMENTED** [review]
Could not get this to work locally yet. I have:

> Having a server.crt and a server.key saved to the base level of your local working directory
> Having a .env.local file defined (see DUOS messages)
> Having env: 'local' in your config.json

And I have a `.env.local` file defined:
```
HOST=local.broadinstitute.org
HTTPS=true
SSL_CRT_FILE=server.crt
SSL_KEY_FILE=server.key
```

The console log errors look like:
```
https://local.broadinstitute.org:3000/api/collections/role/admin 504 (Gateway Timeout)
```

**Comment by rushtong** [comment]
> Could not get this to work locally yet

Update: This is working for me now ... shutting down and restarting seems to have fixed whatever I was doing wrong.

**Review by rushtong — APPROVED** [review]
Definitely working and a solid improvement. We could use more docs on set up in the readme and `setupProxy.js`, but that's good for a follow-on PR.

**Comment by connorlbark** [comment]
I have a feeling this is an issue with my setup, but I cannot seem to get this to work. I can't get mac os to accept the crt/key for some reason but if I force it to go through, Google gives me a `idpiframe_initialization_failed` error.

```
Not a valid origin for the client: https://local.broadinstitute.org:3000 has not been registered for client ID <id>.apps.googleusercontent.com. Please go to https://console.developers.google.com/ and register this origin for your project's client ID.

```

**Review by connorlbark — APPROVED** [review]
with the new client id, works perfectly!
